// tsimp/dist/commonjs/equivalents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.equivalents = exports.isExt = exports.allExts = exports.map = exports.isJSExt = exports.isTSExt = exports.tsExts = exports.jsExts = void 0;
const path_1 = require("path");
exports.jsExts = ['.js', '.mjs', '.cjs', '.jsx'];
exports.tsExts = ['.ts', '.mts', '.cts', '.tsx'];
const isTSExt = (e) => exports.tsExts.includes(e);
exports.isTSExt = isTSExt;
const isJSExt = (e) => exports.jsExts.includes(e);
exports.isJSExt = isJSExt;
exports.map = {
    '.js': ['.ts', '.tsx'],
    '.mjs': ['.mts'],
    '.cjs': ['.cts'],
    '.jsx': ['.tsx'],
    '.ts': ['.js'],
    '.mts': ['.mjs'],
    '.cts': ['.cjs'],
    '.tsx': ['.js', '.jsx'],
};
exports.allExts = Object.keys(exports.map);
const isExt = (e) => exports.allExts.includes(e);
exports.isExt = isExt;
const equivalents = (path, extensionless = false) => {
    const { ext } = (0, path_1.parse)(path);
    const stem = path.substring(0, path.length - ext.length);
    const equivs = [];
    if ((0, exports.isExt)(ext)) {
        equivs.push(...exports.map[ext].map(e => stem + e));
    }
    if (extensionless) {
        equivs.push(...exports.allExts.map(e => path + e));
    }
    return equivs;
};
exports.equivalents = equivalents;
