// socket-post-message/dist/esm/index.js
import EventEmitter from 'node:events';
import { message, Reader } from './message.js';
export { message, Reader };
export class MessageHost extends EventEmitter {
    #stream;
    #reader;
    #writable = true;
    constructor(stream) {
        super();
        this.#stream = stream;
        this.#reader = new Reader();
        this.#stream.on('data', chunk => this.#onData(chunk));
        this.#stream.on('end', () => (this.#writable = false));
        this.#stream.on('close', () => (this.#writable = false));
        this.#stream.on('error', (er) => this.emit('error', er));
    }
    get writable() {
        return this.#writable;
    }
    end() {
        this.#writable = false;
        this.#stream.end();
    }
    postMessage(msg) {
        if (!this.#writable) {
            throw new Error('cannot postMessage after stream end');
        }
        let header;
        let body;
        try {
            ;
            [header, body] = message(msg);
        }
        catch (er) {
            this.emit('error', er);
            return;
        }
        this.#stream.write(header);
        this.#stream.write(body);
    }
    #onData(chunk) {
        let msgs;
        try {
            msgs = this.#reader.write(chunk);
        }
        catch (er) {
            return this.emit('error', er);
        }
        for (const msg of msgs) {
            this.emit('message', msg);
        }
    }
    emit(event, ...data) {
        return super.emit(event, ...data);
    }
    on(event, handler) {
        return super.on(event, handler);
    }
}
export const socketPostMessage = (socket) => new MessageHost(socket);
