// sock-daemon/dist/commonjs/ping.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPing = exports.isPong = exports.ping = exports.pong = void 0;
const pong = (ping) => ({
    ...ping,
    PING: 'PONG',
    pid: process.pid,
});
exports.pong = pong;
function ping(id) {
    return {
        ...(id && { id }),
        PING: 'PING',
        sent: performance.now(),
    };
}
exports.ping = ping;
const isPong = (msg, ping) => !!msg &&
    typeof msg === 'object' &&
    !Array.isArray(msg) &&
    (ping === undefined
        ? typeof msg.id === 'string'
        : msg.id === ping.id) &&
    (ping === undefined
        ? typeof msg.sent === 'number'
        : msg.sent === ping.sent) &&
    msg.PING === 'PONG' &&
    typeof msg.pid === 'number' &&
    Object.keys(msg).length === 4;
exports.isPong = isPong;
const isPing = (msg) => !!msg &&
    typeof msg === 'object' &&
    !Array.isArray(msg) &&
    typeof msg.id === 'string' &&
    msg.PING === 'PING' &&
    typeof msg.sent === 'number' &&
    Object.keys(msg).length === 3;
exports.isPing = isPing;
